src/java.desktop/share/classes/java/awt/GraphicsDevice.java

```diff
@@ -303,23 +296,15 @@ public void setFullScreenWindow(Window w) {
             fullScreenWindow.setBounds(windowedModeBounds);
         }
         // Set the full screen window
-        synchronized (fsAppContextLock) {
```

Comment:
A `synchronized` block has stronger memory consistency guarantees than a `volatile` modifier.

Did you consider leaving the synchronized block but removing the usage of `AppContext`?

Reply:
What do you mean by stronger in this use case?

Reply:
> I mean that there's a _happens before_ relation each time the synchronized block is reach whereas with volatile, the _happens before_ relation is guaranteed only when the value of the variable changes.

I do not think that this is true "guaranteed only when the value of the variable changes".
>A write to a volatile field ([§8.3.1.4](https://docs.oracle.com/javase/specs/jls/se14/html/jls-8.html#jls-8.3.1.4)) happens-before every subsequent read of that field.

Reply:
What I mean is two threads that called `GraphicsDevice.getFullScreenWindow` still established a *happens-before` relation for each call, which will be gone with the volatile modifier.

It may be fine… or may be not… this is why I'm asking.

Reply:
>What I mean is two threads that called GraphicsDevice.getFullScreenWindow still established a happens-before relation for each call, which will be gone with the volatile modifier.
>It may be fine… or may be not… this is why I'm asking.

That only affects the case where the second thread tries to read shared data (unrelated to GraphicsDevice) without synchronization, which was written before the first thread called getFullScreenWindow(). If visibility is required, it should be ensured by the caller. Depending on some internal and undocumented lock for that is not a good thing.

Reply:
Yes, I agree race conditions and memory inconsistencies are still in place, however, the synchronized block brought the state of the full screen object to a fresher state.

I looked through the usages of `getFullScreenWindow()` and couldn't find any usages of full screen window fields, so a `volatile` modifier seems appropriate.